src/packages/hex.gleam

```diff
@@ -3,41 +3,204 @@ import gleam/json
 import gleam/http
 import gleam/result
 import gleam/hackney
+import gleam/list
 import gleam/dynamic.{Decoder} as d
+import time/parse.{parse_iso8601_to_gregorian_seconds}
+import gleam/pgo
+import gleam/string
+import gleam/int
+import gleam/option.{Option}
 
 pub type Error {
   HackneyError(hackney.Error)
   JsonError(json.DecodeError)
+  DatabaseError(pgo.QueryError)
+  ItemNotInListError(Nil)
 }
 
-pub fn query() {
+pub fn query(db: pgo.Connection) {
+  try last_scanned = get_last_scanned(db)
+
+  try packages = query_all_packages([], 1, last_scanned)
+
+  // Only update last scanned at if packages were actually
+  // found that way we only ever use the time from
+  // Hex's API 
+  case packages
+  |> list.length >= 1 {
+    True -> {
+      let len =
+        packages
+        |> list.length
+      assert Ok(last_package) =
+        packages
+        |> list.at(len - 1)
+        |> result.map_error(ItemNotInListError)
+      assert Ok(_) =
+        update_last_scanned(db, parse.parse_iso8601(last_package.updated_at))
+      Ok(Nil)
+    }
+    _ -> Ok(Nil)
+  }
+
+  packages
+  |> list.filter_map(sort_packages)
+  |> io.debug
+  |> Ok
+}
+
+const last_scanned_query = "SELECT date_trunc('second', scanned_at) FROM previous_hex_api_scan LIMIT 1;"
+
+fn get_last_scanned(db: pgo.Connection) -> Result(Int, Error) {
+  let ints = d.tuple3(d.int, d.int, d.int)
+  let datetime_decoder = d.tuple2(ints, ints)
+
+  try response =
+    pgo.execute(last_scanned_query, db, [], d.element(0, datetime_decoder))
+    |> result.map_error(DatabaseError)
+
+  case response.rows {
+    [first, ..] -> Ok(parse.to_gregorian_seconds(first))
+    _ -> Ok(0)
+  }
+}
+
+fn update_last_scanned(
+  db: pgo.Connection,
+  date: parse.LocalDateTime,
+) -> Result(pgo.Returned(d.Dynamic), Error) {
+  // It is done this way as I could not figure out how to
+  // inject a time as a string with the built in $1 system but
+  // this is safe as the value is constructed from a known datatype
```

Comment:
We never build SQL with strings, it is extremely dangerous. Let's pass in the date. I suspect we can give it in iso string format

Reply:
I tried passing it as an iso string, and as a pgo.text with the `pg_time` function and it threw an unexpected type error. Maybe pgo needs a timestamp value?

Reply:
Looks like it https://github.com/erleans/pgo/blob/83ac274e33ba602dd46a07e4c4ed5b7828b24b44/test/pgo_datetime_SUITE.erl#L78

We'll improve the Gleam pgo library to make this nicer in future